File: packages/ios/Sources/Components/Stepper/Stepper.swift
import SwiftUI

public struct KozmosStepper: View {
    let steps: [String]
    let currentStep: Int
    
    public init(steps: [String], currentStep: Int) {
        self.steps = steps
        self.currentStep = currentStep
    }
    
    public var body: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing0) {
            ForEach(0..<steps.count, id: \.self) { index in
                VStack(spacing: KozmosDimensions.primitivesLayoutSpacing50) {
                    // A 32 circle, as Figma and React draw it, its ring inside
                    // the circle: 2 for the current and completed steps, 1 for
                    // a pending one, as the plugin paints them. It was 24 with
                    // a centred 2 on every step until 2026-09-22. The accent is
                    // React's primary pair: theme/600 with foreground/1000 on
                    // it, and the current number in ink. It was theme/500 with
                    // background/0, black on the saturated blue in the dark.
                    ZStack {
                        Circle()
                            .strokeBorder(index <= currentStep ? KozmosColors.primitivesColorsTheme600 : KozmosColors.primitivesColorsForeground500, lineWidth: index <= currentStep ? 2 : 1)
                            .background(Circle().fill(index < currentStep ? KozmosColors.primitivesColorsTheme600 : Color.clear))
                            .frame(width: KozmosDimensions.primitivesLayoutSizing400, height: KozmosDimensions.primitivesLayoutSizing400)
                        
                        if index < currentStep {
                            Image(systemName: "checkmark")
                                .font(KozmosTypography.caption)
                                .fontWeight(.bold)
                                .foregroundColor(KozmosColors.primitivesColorsForeground1000)
                        } else {
                            Text("\(index + 1)")
                                .font(KozmosTypography.caption)
                                .foregroundColor(index == currentStep ? KozmosColors.primitivesColorsForeground0 : KozmosColors.primitivesColorsForeground400)
                        }
                    }
                    
                    // React's label: the current step's in the foreground at medium
                    // weight, every other in the muted foreground, foreground/400
                    // (4.8:1 even on a sheet's grey). It was foreground/100 and /500
                    // until 2026-09-22.
                    Text(steps[index])
                        .font(KozmosTypography.caption)
                        .fontWeight(index == currentStep ? .medium : .regular)
                        .foregroundColor(index == currentStep ? KozmosColors.primitivesColorsForeground0 : KozmosColors.primitivesColorsForeground400)
                }
                
                if index < steps.count - 1 {
                    Rectangle()
                        .fill(index < currentStep ? KozmosColors.primitivesColorsTheme600 : KozmosColors.semanticsBorderSubtle)
                        .frame(height: 1)
                        .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing50)
                }
            }
        }
    }
}

File: packages/android/src/main/java/com/kozmos/components/Stepper/Stepper.kt
package com.kozmos.components.stepper

import com.kozmos.tokens.KozmosDimensions

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosThemeTokens

@Composable
fun KozmosStepper(
    steps: List<String>,
    currentStep: Int,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        steps.forEachIndexed { index, step ->
            val isCompleted = index < currentStep
            val isCurrent = index == currentStep
            
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing50)
            ) {
                Box(
                    modifier = Modifier
                        .size(KozmosDimensions.primitivesLayoutSizing400)
                        .clip(CircleShape)
                        // React's primary pair: theme/600 with foreground/1000
                        // on it, the current number in ink. It was theme/500
                        // with background/0: black on the blue in the dark.
                        .background(if (isCompleted) KozmosThemeTokens.primitivesColorsTheme600 else Color.Transparent)
                        // 2 for the current and completed steps, 1 for a
                        // pending one, as the plugin paints them and React
                        // draws them. Every step was 2 until 2026-09-22.
                        .border(
                            width = if (isCompleted || isCurrent) KozmosDimensions.primitivesLayoutSpacing25 else 1.dp,
                            color = if (isCompleted || isCurrent) KozmosThemeTokens.primitivesColorsTheme600 else KozmosThemeTokens.primitivesColorsForeground500,
                            shape = CircleShape
                        ),
                    contentAlignment = Alignment.Center
                ) {
                    if (isCompleted) {
                        Icon(
                            imageVector = Icons.Default.Check,
                            contentDescription = null,
                            tint = KozmosThemeTokens.primitivesColorsForeground1000,
                            modifier = Modifier.size(KozmosDimensions.primitivesLayoutSizing200)
                        )
                    } else {
                        Text(
                            text = (index + 1).toString(),
                            style = MaterialTheme.typography.labelMedium,
                            color = if (isCurrent) KozmosThemeTokens.primitivesColorsForeground0 else KozmosThemeTokens.primitivesColorsForeground400
                        )
                    }
                }
                
                // React's label: the current step's in the foreground at medium
                // weight, every other in the muted foreground, foreground/400
                // (4.8:1 even on a sheet's grey). It was foreground/100 and /500
                // until 2026-09-22.
                Text(
                    text = step,
                    style = MaterialTheme.typography.labelSmall,
                    fontWeight = if (isCurrent) FontWeight.Medium else FontWeight.Normal,
                    color = if (isCurrent) KozmosThemeTokens.primitivesColorsForeground0 else KozmosThemeTokens.primitivesColorsForeground400
                )
            }
            
            if (index < steps.size - 1) {
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .height(KozmosDimensions.primitivesLayoutSpacing25)
                        .background(if (index < currentStep) KozmosThemeTokens.primitivesColorsTheme600 else KozmosThemeTokens.semanticsBorderSubtle)
                )
            }
        }
    }
}
